File: java/ru/eustas/twim/Range.java
package ru.eustas.twim;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;

final class Range {

  // Collection of classes. Should not be instantiated.
  Range() {}

  private static final int NUM_NIBBLES = 6;
  private static final int NIBBLE_BITS = 8;
  private static final int NIBBLE_MASK = (1 << NIBBLE_BITS) - 1;
  private static final int VALUE_BITS = NUM_NIBBLES * NIBBLE_BITS;
  private static final long VALUE_MASK = (1L << VALUE_BITS) - 1L;
  private static final int HEAD_NIBBLE_SHIFT = VALUE_BITS - NIBBLE_BITS;
  private static final long HEAD_START = 1L << HEAD_NIBBLE_SHIFT;
  private static final int RANGE_LIMIT_BITS = HEAD_NIBBLE_SHIFT - NIBBLE_BITS;
  private static final long RANGE_LIMIT_MASK = (1L << RANGE_LIMIT_BITS) - 1L;

  static final class Decoder {
    private long low;
    private long range = VALUE_MASK;
    private long code;
    private byte[] data;
    private int offset;

    Decoder(byte[] data) {
      this.data = data;
      long code = 0;
      for (int i = 0; i < NUM_NIBBLES; ++i) {
        code = (code << NIBBLE_BITS) | readNibble();
      }
      this.code = code;
    }

    private int readNibble() {
      return (offset < data.length) ? (data[offset++] & NIBBLE_MASK) : 0;
    }

    final void removeRange(int bottom, int top) {
      low += bottom * range;
      range *= top - bottom;
      while (true) {
        if ((low ^ (low + range - 1)) >= HEAD_START) {
          if (range > RANGE_LIMIT_MASK) {
            break;
          }
          range = -low & RANGE_LIMIT_MASK;
        }
        code = ((code << NIBBLE_BITS) & VALUE_MASK) | readNibble();
        range = (range << NIBBLE_BITS) & VALUE_MASK;
        low = (low << NIBBLE_BITS) & VALUE_MASK;
      }
    }

    final int currentCount(int totalRange) {
      range /= totalRange;
      return (int)((code - low) / range);
    }
  }

  static final class Encoder {
    private long low;
    private long range = VALUE_MASK;
    private ByteArrayOutputStream data = new ByteArrayOutputStream();

    final void encodeRange(int bottom, int top, int totalRange) {
      range /= totalRange;
      low += bottom * range;
      range *= top - bottom;
      while (true) {
        if ((low ^ (low + range - 1)) >= HEAD_START) {
          if (range > RANGE_LIMIT_MASK) {
            break;
          }
          range = -low & RANGE_LIMIT_MASK;
        }
        data.write((byte) (low >> HEAD_NIBBLE_SHIFT));
        range = (range << NIBBLE_BITS) & VALUE_MASK;
        low = (low << NIBBLE_BITS) & VALUE_MASK;
      }
    }

    final byte[] finish() {
      for (int i = 0; i < NUM_NIBBLES; ++i) {
        data.write((byte)(low >> HEAD_NIBBLE_SHIFT));
        low = (low << NIBBLE_BITS) & VALUE_MASK;
      }
      return data.toByteArray();
    }
  }

  static final class OptimalEncoder {
    Encoder encoder = new Encoder();
    List<Integer> raw = new ArrayList<Integer>();

    final void encodeRange(int bottom, int top, int totalRange) {
      raw.add(bottom);
      raw.add(top);
      raw.add(totalRange);
      encoder.encodeRange(bottom, top, totalRange);
    }

    private static int checkData(byte[] data, List<Integer> raw) {
      Decoder d = new Decoder(data);
      for (int i = 0; i < raw.size(); i += 3) {
        int bottom = raw.get(i);
        int top = raw.get(i + 1);
        int totalRange = raw.get(i + 2);
        int c = d.currentCount(totalRange);
        if (c < bottom) {
          return -1;
        } else if (c >= top) {
          return 1;
        }
        d.removeRange(bottom, top);
      }
      return 0;
    }

    final byte[] finish() {
      byte[] data = this.encoder.finish();
      int dataLength = data.length;
      while (dataLength > 0) {
        byte lastByte = data[--dataLength];
        data[dataLength] = 0;
        if (lastByte == 0 || checkData(data, raw) == 0) {
          continue;
        }
        int offset = dataLength - 1;
        while (offset >= 0 && data[offset] == -1) {
          offset--;
        }
        if (offset < 0) {
          data[dataLength++] = lastByte;
          break;
        }
        data[offset]++;
        for (int i = offset + 1; i < dataLength; ++i) {
          data[i] = 0;
        }
        if (checkData(data, this.raw) == 0) {
          continue;
        }
        data[offset]--;
        for (int i = offset + 1; i < dataLength; ++i) {
          data[i] = -1;
        }
        data[dataLength++] = lastByte;
        break;
      }
      byte[] result = new byte[dataLength];
      System.arraycopy(data, 0, result, 0, dataLength);
      return result;
    }
  }
}

File: js/range.js
// NUM_NIBBLES = 6
// NIBBLE_BITS = 8
// NIBBLE_MUL = 256 = (1 << NIBBLE_BITS)
// NIBBLE_MASK = 255 = (1 << NIBBLE_BITS) - 1
// VALUE_BITS = 48 = NUM_NIBBLES * NIBBLE_BITS
// VALUE_MUL = 281474976710656 = (1 << VALUE_BITS)
// VALUE_MASK = 281474976710655 = (1 << VALUE_BITS) - 1
// HEAD_NIBBLE_SHIFT = 40 = VALUE_BITS - NIBBLE_BITS
// HEAD_START = 1099511627776 = 1 << HEAD_NIBBLE_SHIFT
// RANGE_LIMIT_BITS = 32 = HEAD_NIBBLE_SHIFT - NIBBLE_BITS
// RANGE_LIMIT_MASK = 4294967295 = (1 << RANGE_LIMIT_BITS) - 1

/**
 * @param{!Array<number>} data encoded bytes
 * @param{!Array<number>} raw flattened original triplets
 * @return{number}
 */
function checkData(data, raw) {
  let d = new Decoder(data);
  for (let i = 0; i < raw.length; i += 3) {
    let c = d.currentCount(raw[i + 2]);
    if (c < raw[i]) {
      return -1;
    } else if (c >= raw[i + 1]) {
      return 1;
    }
    d.removeRange(raw[i], raw[i + 1]);
  }
  return 0;
}

class Decoder {
  /**
   * @param{!Array<number>} data bytes
   */
  constructor(data) {
    /** @private{number} */
    this.low = 0;
    /** @private{number} */
    this.range = 281474976710655;
    /** @const @private{!Array<number>} */
    this.data = data;
    /** @private{number} */
    this.offset = 0;
    /** @private{number} */
    this.code = 0;
    for (let i = 0; i < 6; ++i) {
      this.code = (this.code * 256) + (this.data[this.offset++] | 0);
    }
  }

  /**
   * @param{number} bottom
   * @param{number} top
   * @return {void}
   */
  removeRange(bottom, top) {
    this.low += bottom * this.range;
    this.range *= top - bottom;
    while (true) {
      if (Math.floor(this.low / 1099511627776) !== Math.floor((this.low + this.range - 1) / 1099511627776)) {
        if (this.range > 4294967295) {
          break;
        }
        this.range = (281474976710656 - this.low) % 4294967296;
      }
      this.code = ((this.code % 1099511627776) * 256) + (this.data[this.offset++] | 0);
      this.range = (this.range % 1099511627776) * 256;
      this.low = (this.low % 1099511627776) * 256;
    }
  }

  /**
   * @param{number} totalRange
   * @return{number}
   */
  currentCount(totalRange) {
    this.range = Math.floor(this.range / totalRange);
    return Math.floor((this.code - this.low) / this.range);
  }
}

class Encoder {
  constructor() {
    /** @private{number} */
    this.low = 0;
    /** @private{number} */
    this.range = 281474976710655;
    /** @private{!Array<number>} */
    this.data = [];
  }

  /**
   * @param{number} bottom
   * @param{number} top
   * @param{number} totalRange
   * @return {void}
   */
  encodeRange(bottom, top, totalRange) {
    this.range = Math.floor(this.range / totalRange);
    this.low += bottom * this.range;
    this.range *= top - bottom;
    while (true) {
      if (Math.floor(this.low / 1099511627776) !== Math.floor((this.low + this.range - 1) / 1099511627776)) {
        if (this.range > 4294967295) {
          break;
        }
        this.range = (281474976710656 - this.low) % 4294967296;
      }
      this.data.push(Math.floor(this.low / 1099511627776) & 255);
      this.range = (this.range % 1099511627776) * 256;
      this.low = (this.low % 1099511627776) * 256;
    }
  }

  /**
   * @return{!Array<number>}
   */
  finish() {
    for (let i = 0; i < 6; ++i) {
      this.data.push(Math.floor(this.low / 1099511627776) & 255);
      this.low = (this.low % 1099511627776) * 256;
    }
    return this.data;
  }
}

class OptimalEncoder {
  constructor() {
    /** @const @private{!Encoder} */
    this.encoder = new Encoder();
    /** @const @private{!Array<number>} */
    this.raw = [];
  }

  /**
   * @param{number} bottom
   * @param{number} top
   * @param{number} totalRange
   * @return {void}
   */
  encodeRange(bottom, top, totalRange) {
    this.raw.push(bottom, top, totalRange);
    this.encoder.encodeRange(bottom, top, totalRange);
  }

  /**
   * @return{!Array<number>}
   */
  finish() {
    let data = this.encoder.finish();
    while (true) {
      if (data.length === 0) {
        return data;
      }
      let lastByte = data.pop();
      if ((lastByte === 0) || (checkData(data, this.raw) === 0)) {
        continue;
      }
      let offset = data.length - 1;
      while (offset >= 0 && data[offset] === 255) {
        offset--;
      }
      if (offset < 0) {
        data.push(lastByte);
        return data;
      }
      data[offset]++;
      for (let i = offset + 1; i < data.length; ++i) {
        data[i] = 0;
      }
      if (checkData(data, this.raw) === 0) continue;
      data[offset]--;
      for (let i = offset + 1; i < data.length; ++i) {
        data[i] = 255;
      }
      data.push(lastByte);
      return data;
    }
  }
}

/** @export */
const range = {};
range.Decoder = Decoder;
range.Encoder = Encoder;
range.OptimalEncoder = OptimalEncoder;
